src/rshim.c

```diff
@@ -2157,30 +2195,152 @@ static int rshim_update_locked_mode(rshim_backend_t *bd)
       int rt;
 
       rshim_lock();
-      pthread_mutex_lock(&bd->mutex);
       rt = rshim_access_check(bd); 
-      pthread_mutex_unlock(&bd->mutex);
       rshim_unlock();
       if (rt) {
         RSHIM_INFO("rshim%d attached by another device. Entering Drop Mode\n",
             bd->index);
+        pthread_mutex_lock(&bd->mutex);
         rshim_set_drop_mode(bd, 1);
+        pthread_mutex_unlock(&bd->mutex);
       }
     }
   }
 
   return 0;
 }
 
+static int rshim_check_sp1_magic(rshim_backend_t *bd, uint64_t magic,
+    bool* result)
+{
+  int rc;
+  uint64_t sp1;
+
+  rc = bd->read_rshim(bd, RSHIM_CHANNEL, bd->regs->scratchpad1, &sp1, RSHIM_REG_SIZE_8B);
+  if (rc || RSHIM_BAD_CTRL_REG(sp1)) {
+    RSHIM_ERR("rshim%d failed to read sp1: %d\n", bd->index, rc);
+    *result = false;
+    return rc;
+  }
+
+  *result = (sp1 == magic);
+
+  return 0;
+}
+
+static int rshim_write_sp1_magic(rshim_backend_t *bd, uint64_t magic)
+{
+  int rc;
+
+  RSHIM_INFO("rshim%d writing sp1 magic %s (0x%llx)\n", bd->index,
+      magic_to_str(magic), (unsigned long long)magic);
+
+  rc = bd->write_rshim(bd, RSHIM_CHANNEL, bd->regs->scratchpad1, magic,
+      RSHIM_REG_SIZE_8B);
+  if (rc) {
+    RSHIM_ERR("rshim%d failed to write sp1: %s\n", bd->index, strerror(-rc));
+    return -ENODEV;
+  }
+
+  return 0;
+}
+
+static int rshim_handle_ownership_transfer(rshim_backend_t *bd)
+{
+  int i, rt;
+  bool has_ack;
+  bool has_req;
+
+  if (bd->drop_mode) {
+    if (bd->force_cmd_pending) {
+      RSHIM_INFO("rshim%d executing Force command\n", bd->index);
+      bd->force_cmd_pending = 0;
+
+      if (bd->enable_device && bd->enable_device(bd, true)) {
+        RSHIM_ERR("rshim%d failed to enable device\n", bd->index);
+        return -EIO;
+      }
+
+      bd->requesting_rshim = 1;
+
+      /* sending req and checking ack multiple times to ensure the transfer */
+      for (i = 0; i < 10; i++) {
+        rt = rshim_write_sp1_magic(bd, RSHIM_OSP_REQ_MAGIC_NUM);
+        if (rt) {
+          RSHIM_ERR("rshim%d failed to write ownership transfer req\n",
+              bd->index);
+          continue;
+        }
+
+        usleep(RSHIM_OSP_TO_REQ_MS * 1000 / 10);
+
+        has_ack = false;
+        rt = rshim_check_sp1_magic(bd, RSHIM_OSP_ACK_MAGIC_NUM, &has_ack);
+
+        if (!rt && has_ack) {
+          break;
+        }
+      }
+
+      if (rt || !has_ack) {
+        RSHIM_WARN("rshim%d failed to receive ownership transfer ack\n",
+            bd->index);
+        bd->requesting_rshim = 0;
+        return -EIO;
+      }
+      RSHIM_INFO("rshim%d received ownership transfer ack\n", bd->index);
+
+      rt = rshim_set_drop_mode(bd, 0);
+      if (rt) {
```

Comment:
if (rt || bd->drop_mode)

Reply:
Not sure why `|| bd->drop_mode` is needed. If "rt" is non-zero, meaning the function failed to set `bd->drop_mode` from 1 to 0, which means it must be `bd->drop_mode` must be 1 already.

Reply:
Yes, you're right. For this case, checking 'rt' should be good enough though checking both might be more robust (for example, if drop_mode is already 0 and we call rshim_set_drop_mode(bd, 0) again, it'll return failure. Up to you whether you want to add this change.

Reply:
Then should it be `rt && bd->drop_mode` here?

Reply:
Or how about `if (rt && rt != -EALREADY)`?

Reply:
Suggest either "if (rt)" like you did before,  or "if (bd->drop_mode)"

Reply:
Use `if(rt)`, but make rshim_set_drop_mode() to return 0 when the value is already there: 1e89c98.